### PR TITLE
Fix return for virtual product 

### LINKED ISSUE #16275: Prevent customers from returning a full virtual order
**Is your feature request related to a problem? Please describe:**

When an order contains only virtual products, the Merchandise Returns tab displays the following message: Merchandise returns are not applicable for virtual orders. So we let the merchant know virtual products (and by extension, fully virtual orders) cannot be returned.

But once a virtual order is noted as delivered, the customer can connect to his/her account, go in the details of his/her virtual order, select all virtual products and decide to return them.

Back from a merchant's side, the Merchandise Returns tab remains the same, with the message displayed and the customer return not listed.

**Describe the solution you'd like:**

It would make sense to prevent the customer from returning his/her virtual order. Perhaps by displaying the same message?

cf. https://github.com/PrestaShop/PrestaShop/pull/16084#discussion_r338060225

### PR DESCRIPTION
**Pull request by ghost**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Customer can return virtual product this PR allow to return only no virtual product.
| Type?             | improvement
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/16275
| How to test?      |  2 steps : Made an order with only virtual product and check return Made an order with a virtual product and a no virtual product check the return. Order need to be shipped and return enable in the back office
| Possible impacts? | No

Order with one virtual product and one no virtual :

Capture d’écran de 2021-07-01 16-42-40

Order with only virtual product :

Capture d’écran de 2021-07-01 16-43-56

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/25165)

This PR Close : https://github.com/PrestaShop/PrestaShop/pull/25165/files ( noob rebase )

### CODE REVIEW

**Comment by Julievrz** [comment]
Hi, I'm fine with the wording. I'm just not sure it should be placed here on the page. @prestascott, what do you think?

**Comment by ghost** [comment]
@prestascott  @Julievrz to complete we can also have two more information here Download Link and Customization Link 

124145881-7c902f00-da8d-11eb-953d-43d2fba6e018

**Comment by prestascott** [comment]
Hello everyone, 

Thanks @okom3pom for your PR. But why your screen is in french and english at the same time? 🤔 
Imo, the "Virtual products can't be returned." message needs to be more visible. We can change its color to orange for example.

Please correct me if I'm wrong but the "téléchargement" button allows customers to download the virtual product right?

**Comment by ghost** [comment]
Hello @prestascott 

> Thanks @okom3pom for your PR. But why your screen is in french and english at the same time? :thinking: 

It's a new wording so we have english and french

> Imo, the "Virtual products can't be returned." message needs to be more visible. We can change its color to orange for example.

Why not but i think @Julievrz want to change the position ?

> Please correct me if I'm wrong but the "téléchargement" button allows customers to download the virtual product right? 

Yes, but you can also have at the same time: the download and the personalization.

**Comment by ghost** [comment]
A screenshot with the full page plz say me where i put the message and if color is needed ?

I think it's better to se with the full page :) 

Regards

screencapture-127-0-0-1-dev-PrestaShop-index-php-2021-07-05-16_28_05

**Comment by ghost** [comment]
@prestascott any news ?

**Comment by prestascott** [comment]
Hey @okom3pom,

I would simply change the color of the warning message and put the "download" and "customization" links next to each other. 
Also, I would change the wording of the button to "Download the virtual product" to make the action more explicit.

Wdyt? @Julievrz 

What does the "customization" link allow you to do? @okom3pom 
Maybe we can change the wording too :)

Here is my design proposition:
Frame 13

**Comment by ghost** [comment]
`What does the "customization" link allow you to do? `

It's a Pop-up with your cutomization

**Comment by Julievrz** [comment]
> Hey @okom3pom,
> 
> I would simply change the color of the warning message and put the "download" and "customization" links next to each other.
> Also, I would change the wording of the button to "Download the virtual product" to make the action more explicit.
> 
> Wdyt? @Julievrz
> 
> What does the "customization" link allow you to do? @okom3pom
> Maybe we can change the wording too :)
> 
> Here is my design proposition:
> Frame 13

Hi @prestascott @okom3pom, sorry for the late answer, I was out of the office. I'm fine with replacing 'Download' with 'Download virtual product' (without the "the") but I'm afraid it's out of the scope of this issue. If @okom3pom is ok to fix it, then, why not! :-) In that case, we can take the opportunity to replace "Customization" with "See customization" as well, for more consistency.

**Comment by ghost** [comment]
Hello,

I am ok to change the Wording and put : Download virtual product and See customization in one line. For the orange i open a new issue and a new PR because i am bad with css i need help.

**Comment by Julievrz** [comment]
@okom3pom Thank you!

**Comment by Progi1984** [comment]
@okom3pom It seems that this PR needs a rebase ;)

**Review by Progi1984 — DISMISSED** [review]

File: themes/classic/templates/customer/_partials/order-detail-return.tpl [hunk 1]
```diff
@@ -40,23 +40,28 @@
...
-                  <input type="checkbox" id="cb_{$product.id_order_detail}" name="customization_ids[{$product.id_order_detail}][]" value="{$customization.id_customization}">
+            {if !$product.is_virtual}
+                {if !$product.customizations}
+                  <span id="_desktop_product_line_{$product.id_order_detail}">
+                  <input type="checkbox" id="cb_{$product.id_order_detail}" name="ids_order_detail[{$product.id_order_detail}]" value="{$product.id_order_detail}">
                 </span>
-                {/foreach}
-              {/if}
+                {else}
+                  {foreach $product.customizations  as $customization}
```
Comment (sowbiba) [inline themes/classic/templates/customer/_partials/order-detail-return.tpl]: extra space before the 'as'

**Review by sowbiba — DISMISSED** [review]

**Comment by Progi1984** [comment]
@okom3pom Could you rebase your PR ? After that, we can send your PR in QA.

**Comment by ghost** [comment]
Monday :eyes:

**Review by Progi1984 — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by ghost** [comment]
@Julievrz can you confirm the wording :) 

regards Thomas

**Review by matthieu-rolland — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Comment by florine2623** [comment]
Hello all,

Thanks for the PR @okom3pom !

Only virtual products : OK
Screenshot 2021-10-28 at 12 02 17

Mix of virtual and non-virtual products : OK
Screenshot 2021-10-28 at 12 06 53

Switched a Standard product to a Virtual product after customer's order : OK
Screenshot 2021-10-28 at 12 11 38

>Download virtual product and See customization in one line

I don't see this behavior on my end.
I have a Virtual product with customization, both links are on different line :
Screenshot 2021-10-28 at 12 22 48
Could you check ?

Also, this might be out-of-scope, but for the same product, sometimes the direct link to the product is available and sometimes not available :
Screenshot 2021-10-28 at 12 28 43
Screenshot 2021-10-28 at 12 28 48
If it is out-of-scope, I can't create an issue :)

Thanks!

**Comment by ghost** [comment]
Hello,

For the direct link it's out of scope.

For 
139238053-c9ca2d20-b4c2-4551-b2b9-f3898dd4971a

I will put the message after 

Thank you for reviews it was not easy

**Comment by Progi1984** [comment]
@okom3pom Hi, Could you rebase, please ?

File: themes/classic/templates/customer/_partials/order-detail-return.tpl [hunk 2]
```diff
@@ -43,6 +43,20 @@
...
                 <input type="checkbox" id="cb_{$product.id_order_detail}" name="ids_order_detail[{$product.id_order_detail}]" value="{$product.id_order_detail}">
               </span>
+
+            {if !$product.is_virtual}
+                {if !$product.customizations}
+                  <span id="_desktop_product_line_{$product.id_order_detail}">
+                  <input type="checkbox" id="cb_{$product.id_order_detail}" name="ids_order_detail[{$product.id_order_detail}]" value="{$product.id_order_detail}">
+                </span>
+                {else}
+                  {foreach $product.customizations  as $customization}
```
Comment (atomiix) [inline themes/classic/templates/customer/_partials/order-detail-return.tpl]:
```suggestion
                  {foreach $product.customizations as $customization}
```

File: themes/classic/templates/customer/_partials/order-detail-return.tpl [hunk 3]
```diff
@@ -43,6 +43,20 @@
               <span id="_desktop_product_line_{$product.id_order_detail}">
                 <input type="checkbox" id="cb_{$product.id_order_detail}" name="ids_order_detail[{$product.id_order_detail}]" value="{$product.id_order_detail}">
               </span>
+
+            {if !$product.is_virtual}
+                {if !$product.customizations}
+                  <span id="_desktop_product_line_{$product.id_order_detail}">
+                  <input type="checkbox" id="cb_{$product.id_order_detail}" name="ids_order_detail[{$product.id_order_detail}]" value="{$product.id_order_detail}">
+                </span>
```
Comment (atomiix) [inline themes/classic/templates/customer/_partials/order-detail-return.tpl]:
```suggestion
                    <input type="checkbox" id="cb_{$product.id_order_detail}" name="ids_order_detail[{$product.id_order_detail}]" value="{$product.id_order_detail}">
                  </span>
```

File: themes/classic/templates/customer/_partials/order-detail-return.tpl [hunk 4]
```diff
@@ -43,6 +43,20 @@
...
+                {else}
+                  {foreach $product.customizations  as $customization}
+                    <span id="_desktop_product_customization_line_{$product.id_order_detail}_{$customization.id_customization}">
+                    <input type="checkbox" id="cb_{$product.id_order_detail}" name="customization_ids[{$product.id_order_detail}][]" value="{$customization.id_customization}">
+                  </span>
```
Comment (atomiix) [inline themes/classic/templates/customer/_partials/order-detail-return.tpl]:
```suggestion
                      <input type="checkbox" id="cb_{$product.id_order_detail}" name="customization_ids[{$product.id_order_detail}][]" value="{$customization.id_customization}">
                    </span>
```

**Review by atomiix — DISMISSED** [review]

**Review by matks — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Comment by ghost** [comment]
@florine2623 I need to open another problem, because I can't put the link in one line. 

It's an other problem we can group with : https://github.com/PrestaShop/PrestaShop/pull/25204#issuecomment-887333924

it's for Front dev :)

**Comment by Progi1984** [comment]
(Ping @NeOMakinG)

**Comment by florine2623** [comment]
Hello @okom3pom ,

I've tested again and I have the same result as before ! So it it OK ✅ 

You can ping me for the new PR with :
- changing "Download" to "Download virtual product"
- changing the color of the text "Virtual products can't be returned." to orange
- changing "Product customization" to "See customization"
- having both links "Download virtual product" & "See customization" on the same line

How it is :
Screenshot 2021-11-16 at 11 01 04

How it should be :
Screenshot 2021-11-16 at 11 03 26

Thanks :)

**Comment by ghost** [comment]
Hello @florine2623 

The exact wording is :  'Download virtual product' (without the "the")  "See customization" @Julievrz [comment](https://github.com/PrestaShop/PrestaShop/pull/25204#issuecomment-887333924)

I open news issue when the PR is  merge

**Comment by Progi1984** [comment]
Thanks @okom3pom & @florine2623

**Pull request merged**